### PR TITLE
Enable marking messages as unread in UI

### LINKED ISSUE #205: Be able to "unread" an email after reading it. [$10]
Some people like to filter their messages this way.

<bountysource-plugin>

---
There is a **[$10 open bounty](https://www.bountysource.com/issues/39507998-be-able-to-unread-an-email-after-reading-it?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by kevgathuku**
Add a dropdown menu with `Delete` and `Toggle read` functionality
Fixes #205 

Preview:
screen shot 2018-07-21 at 19 55 42

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (ChristophWurst) [inline .prettierrc]: please remove this file :)

File: css/mail.scss [hunk 2]
```diff
@@ -522,6 +536,14 @@ input.submit-message,
 	display: inline-block;
 }
 
+.mail-message-actions-menu{
```
Comment (ChristophWurst) [inline css/mail.scss]: cc @skjnldsv regarding the need for this special rule or if we can somehow streamline our open/closed popover states somehow.

File: js/views/messagesitem.js [hunk 3]
```diff
@@ -102,7 +109,16 @@ define(function(require) {
 				revert: 'invalid'
 			});
 
-			tooltip('.action.delete', {placement: 'left'});
+			// TODO: Close mail actions menu on outside click
```
Comment (ChristophWurst) [inline js/views/messagesitem.js]: there's some logic for that. you'll find it at other code sections where we open menus as well. let me know if you can't find it :)

File: js/webpack.dev.config.js [hunk 4]
```diff
@@ -2,5 +2,6 @@ const merge = require('webpack-merge');
 const baseConfig = require('./webpack.base.config.js');
 
 module.exports = merge(baseConfig, {
+	mode: 'development',
```
Comment (ChristophWurst) [inline js/webpack.dev.config.js]: awesome, I forgot to add this here :+1:

File: js/webpack.prod.config.js [hunk 5]
```diff
@@ -3,6 +3,7 @@ const merge = require('webpack-merge');
 const baseConfig = require('./webpack.base.config.js');
 
 module.exports = merge(baseConfig, {
+	mode: 'production',
```
Comment (ChristophWurst) [inline js/webpack.prod.config.js]: same :+1:

File: package-lock.json [hunk 6]
```diff
@@ -332,9 +332,9 @@
       "dev": true
     },
     "agent-base": {
-      "version": "4.2.0",
-      "resolved": "https://registry.npmjs.org/agent-base/-/agent-base-4.2.0.tgz",
-      "integrity": "sha512-c+R/U5X+2zz2+UCrCFv6odQzJdoqI+YecuhnAJLa1zYaMc13zPfwMwZrr91Pd1DYNo/yPRbiM4WVf9whgwFsIg==",
+      "version": "4.2.1",
+      "resolved": "https://registry.npmjs.org/agent-base/-/agent-base-4.2.1.tgz",
+      "integrity": "sha512-JVwXMr9nHYTUXsBFKUqhJwvlcYU/blreOEUkhNR2eXZIvwd+c+o5V4MgDPKWnMS/56awN3TRzIP+KoPn+roQtg==",
```
Comment (ChristophWurst) [inline package-lock.json]: please revert changes to this file, they shouldn't be necessary for this PR.

File: js/views/messagesitem.js (hunk 3 above)
Comment (kevgathuku) [reply js/views/messagesitem.js]: I already found it and implemented it. I just forgot to remove the comment 😅

File: .prettierrc (hunk 1 above)
Comment (kevgathuku) [reply .prettierrc]: 👍

File: package-lock.json (hunk 6 above)
Comment (kevgathuku) [reply package-lock.json]: More specifically, this was as a result of running `npm audit fix`

File: css/mail.scss (hunk 2 above)
Comment (skjnldsv) [reply css/mail.scss]: Yup, @kevgathuku take a look at https://docs.nextcloud.com/server/14/developer_manual/design/list.html :)

File: js/views/messagesitem.js (hunk 3 above)
Comment (ChristophWurst) [reply js/views/messagesitem.js]: hehe, good :grinning:

File: package-lock.json (hunk 6 above)
Comment (ChristophWurst) [reply package-lock.json]: okay, makes sense. I do run these fixes occasionally as well. I'm fine with leaving it this way for this PR, but please just create separate PRs for separate change sets in the future. This makes coordination of PRs easier and also lowers the risk of getting a merge conflict in the package-lock.json file if it takes a bit longer to integrate a feature PR like this :)

Comment (kevgathuku) [reply package-lock.json]: Understood 👍

File: css/mail.scss (hunk 2 above)
Comment (kevgathuku) [reply css/mail.scss]: Thanks for the pointers. This should be fixed now.

**Comment by kevgathuku** [comment]
@ChristophWurst I've fixed all the issues you raised earlier. Ready for you to take another look :)

File: js/views/messagesitem.js [hunk 7]
```diff
@@ -126,42 +146,79 @@ define(function(require) {
...
+			e.preventDefault();
+			this.actionsMenuShown = !this.actionsMenuShown;
+			this.toggleMenuClass();
+		},
+
+		toggleMenuClass: function() {
+			this.getUI('menu').toggleClass('open', this.actionsMenuShown);
+		},
+
+		toggleMessageRead: function(event) {
```
Comment (ChristophWurst) [inline js/views/messagesitem.js]: - [ ] BUG: after clicking `Toggle read`, the menu is neither closed, nor can I close it by clicking the three dots again. IMO I'd go with closing the menu right away after one of the action inside has been clicked.

**Comment by ChristophWurst** [comment]
@skjnldsv somehow the mail app layout is pretty much broken at the moment as the second column after the left sidebar (main content) is displayed *underneath* the sidebar.

Was there any relevant change in our layout structure we have to adjust here for nc14?

File: js/views/messagesitem.js (hunk 7 above)
Comment (kevgathuku) [reply js/views/messagesitem.js]: @ChristophWurst  I unfortunately cannot replicate this on Chrome 67 and Firefox 61
On my end, the menu is closed immediately the `Toggle read` is clicked.

**Comment by skjnldsv** [comment]
@ChristophWurst Yes, the structure has been standardised for 14 as we had a lot of breaking changes for this version. See https://github.com/nextcloud/server/pull/9982

File: js/views/messagesitem.js (hunk 7 above)
Comment (ChristophWurst) [reply js/views/messagesitem.js]: Oh, wait. I think I described the issue wrong. The issue is actually that you can't re-open the menu after one of the actions has been clicked (I guess the toggle flag is missing an update). You have to click the three dots twice to get back into the menu.

Comment (kevgathuku) [reply js/views/messagesitem.js]: Ahh, now I can reproduce the bug :)
I'll take a look at it.
Thanks for the better explanation 👍

Comment (kevgathuku) [reply js/views/messagesitem.js]: @ChristophWurst the bug should be fixed now.

**Comment by ChristophWurst** [comment]
@kevgathuku somehow the three dots have moved to a weird position with the latest commit. They are now centered on the message element and disappears on hover. What version of the Nextcloud server are you using for development?

Since I'm having a hard time reviewing PRs in a state where the page structure is broken (https://github.com/nextcloud/mail/pull/973#issuecomment-407044413), I'll set this PR as `blocked` until we resolved the other issue.

**Comment by ChristophWurst** [comment]
- [ ] Blocked by https://github.com/nextcloud/mail/issues/981

**Comment by kevgathuku** [comment]
Hi @ChristophWurst I wasn't using the latest master because of the same issue. I agree, let's get the page structure fixed then we can get back to this.

**Comment by kevgathuku** [comment]
> Aaaaand we're ready to continue here 😉
Awesome!

I'll look at the PR and fix the conflicts @ChristophWurst

**Comment by kevgathuku** [comment]
Managed to fix the  conflicts. Ready for another look @ChristophWurst

**Comment by ChristophWurst** [comment]
> danger89 approved these changes 2 minutes ago

Did you test the app on this branch and also review the code?

File: Makefile [hunk 8]
```diff
@@ -33,11 +33,14 @@ install-npm-deps:
...
 
 install-npm-deps-dev:
-	npm install --deps
+	npm install
 
 optimize-js: install-npm-deps-dev
 	./node_modules/webpack/bin/webpack.js --config js/webpack.prod.config.js
 
+dev-js: install-npm-deps-dev
+	./node_modules/webpack/bin/webpack.js --config js/webpack.dev.config.js --watch
```
Comment (ChristophWurst) [inline Makefile]: Would you mind changing this to `npm run dev`. It essentially does the same, but prevents duplication of this command. While at it, please also change the `optimize-js` to `npm run build`. Thanks!

Comment (kevgathuku) [reply Makefile]: This should be fixed now @ChristophWurst

**Comment by ChristophWurst** [comment]
@kevgathuku please don't forget to claim your bounty at https://www.bountysource.com/issues/39507998-be-able-to-unread-an-email-after-reading-it :wink:

**Comment by kevgathuku** [comment]
Thanks @ChristophWurst 
Really glad I could help out 😃

**Comment by danger89** [comment]
I would like to say: Keep up the good work! You make the world a better place. Keep creating pull requests like these. Love it :heart:

**Pull request merged**